pkg/conversion/mapi2capi/aws_test.go

```diff
@@ -16,43 +16,140 @@ limitations under the License.
+	}))
+	awsMAPIMachineWithVirtualName = awsMAPIMachineBase.WithProviderSpecBuilder(awsBaseProviderSpec.WithBlockDevices([]mapiv1.BlockDeviceMappingSpec{
+		{VirtualName: ptr.To("test"), EBS: &mapiv1.EBSBlockDeviceSpec{VolumeSize: ptr.To(int64(10))}},
+	}))
+	awsMAPIMachineWithVirtualNameAndRootVolNotDeletedOnTerm = awsMAPIMachineBase.WithProviderSpecBuilder(awsBaseProviderSpec.WithBlockDevices([]mapiv1.BlockDeviceMappingSpec{
+		{VirtualName: ptr.To("test"), EBS: &mapiv1.EBSBlockDeviceSpec{VolumeSize: ptr.To(int64(10)), DeleteOnTermination: ptr.To(false)}},
+	}))
+	awsMAPIMachineWithVirtualNameAndMissingEBSConfig = awsMAPIMachineBase.WithProviderSpecBuilder(awsBaseProviderSpec.WithBlockDevices([]mapiv1.BlockDeviceMappingSpec{
+		{VirtualName: ptr.To("test")},
+	}))
+	awsMAPIMachineWithMissingEBSConfig = awsMAPIMachineBase.WithProviderSpecBuilder(awsBaseProviderSpec.WithBlockDevices([]mapiv1.BlockDeviceMappingSpec{
+		{DeviceName: ptr.To("/dev/sdb")},
+	}))
+	awsMAPIMachineWithRootVolumeNotDeleted = awsMAPIMachineBase.WithProviderSpecBuilder(awsBaseProviderSpec.WithBlockDevices([]mapiv1.BlockDeviceMappingSpec{
+		{EBS: &mapiv1.EBSBlockDeviceSpec{VolumeSize: ptr.To(int64(10)), DeleteOnTermination: ptr.To(false)}},
+	}))
+	awsMAPIMachineWithNoDeviceAndVirtualName = awsMAPIMachineBase.WithProviderSpecBuilder(awsBaseProviderSpec.WithBlockDevices([]mapiv1.BlockDeviceMappingSpec{
+		{VirtualName: ptr.To("test"), NoDevice: ptr.To("test"), EBS: &mapiv1.EBSBlockDeviceSpec{VolumeSize: ptr.To(int64(10))}},
+	}))
+
+	// MachineSets configurations.
+	awsMAPIMachineSetBase = machinebuilder.MachineSet().WithProviderSpecBuilder(awsBaseProviderSpec)
+
+	// OCP Infrastructure configurations.
+	infra = &configv1.Infrastructure{
+		Spec:   configv1.InfrastructureSpec{},
+		Status: configv1.InfrastructureStatus{InfrastructureName: "sample-cluster-name"},
+	}
+	infraWithRegion = configbuilder.Infrastructure().AsAWS("sample-cluster-name", awsDefaultRegion)
 )
 
-var _ = Describe("mapi2capi AWS", func() {
-	awsMAPIMachine := machinebuilder.Machine().WithProviderSpecBuilder(machinebuilder.AWSProviderSpec()).WithProviderID("aws:///us-west-2a/i-05442bc41c3df969d").Build()
-	awsMAPIMachineSet := machinebuilder.MachineSet().WithProviderSpecBuilder(machinebuilder.AWSProviderSpec()).Build()
-	awsMAPIMachineSet.Spec.Template.Spec.ProviderID = ptr.To("aws:///us-west-2a/i-05442bc41c3df969d") // TODO: do this in machinebuilder.
-	infra := configv1.Infrastructure{
-		Spec: configv1.InfrastructureSpec{},
-		Status: configv1.InfrastructureStatus{
-			InfrastructureName: "sample-cluster-name",
-		},
+var _ = DescribeTable("mapi2capi AWS convert MAPI Machine",
+	func(m *mapiv1.Machine, i *configv1.Infrastructure, errorsMatcher types.GomegaMatcher, warningsMatcher types.GomegaMatcher) {
+		_, _, warns, err := FromAWSMachineAndInfra(m, i).ToMachineAndInfrastructureMachine()
+		Expect(err).To(errorsMatcher, "should have been able to convert providerSpec to MachineTemplateSpec")
```

Comment:
You're reducing the value of the table by having the matcher outside of this function, putting too much into the Entry IMO

I would have expected something more akin to
```
if in.expectedError != "" {
  Expect(err).To(MatchError(ContainSubstring(in.expectedError)), "...")
} else {
  Expect(err).To(Not(HaveOcurred()), "...")
}
```

Reply:
Yes, I agree, the matching needs to happen against multiple errors and warnings, so the logic will be slightly different, but let's do this inside the table setup function.

Reply:
Ooh good point, perhaps the expected errors should be a slice? Since we probably want to test that with multiple issues we can see multiple errors

Reply:
Indeed!